fix(button): Allow slotted button content without text

macandcheese opened the pull request:
**Related Issue:** Resolves #1438 cc @MikeTschudi 

## Summary

This re-enables the ability to place things other than text content inside a button's slot.

Because of the way padding is applied to buttons, it means that users may need to manually constrain the height of buttons that contain slotted content other than text. This really shouldn’t be recommended, and the preferred method of adding icons should be icon-start / icon-end, but this change will allow certain unique cases to be achieved. cc @asangma cc @MikeTschudi 

It should be noted Safari can't really handle "user set height and width" like in the added demo. Not sure if @asangma has a neat trick for this, but since this is "non-recommended", I think we can let users handle positioning things inside the buttons with negative margin etc. if needed. I was unable to find a solution to this, it seems like a Safari-only issue (the three buttons in the screenshot's left are using user-set height / width, and custom content - svg or calcite-icon - whereas the fourth example uses icon-start as recommended and displays fine across browsers).

![Slice](https://user-images.githubusercontent.com/4733155/103948457-a6742480-50ee-11eb-8256-6057a37b522a.png)


<!--

Please make sure the PR title and/or commit message adheres to the https://conventionalcommits.org/ specification.

Note: If your PR only has one commit and it is NOT semantic, you will need to either

a. add another commit and wait for the check to update
b. proceed to squash merge, but make sure the commit message is the same as the title.

This is because of the way GitHub handles single-commit squash merges (see https://github.com/zeke/semantic-pull-requests/issues/17)

If this is component-related, please verify that:

- [ ] feature or fix has a corresponding test
- [ ] changes have been tested with demo page in Edge

-->

eriklharper commented on src/components/calcite-button/calcite-button.tsx:
```diff
@@ -138,7 +138,7 @@ export class CalciteButton {
 
     return (
-      <Host dir={dir} hasText={this.hasText}>
+      <Host dir={dir} hasContent={this.hasContent}>
```
Why not just set this prop to `reflect: true`?

driskull replied:
I'm confused on why this is being set on the dom node? A css class should be used on an internal node instead of an invalid undeclared prop.

driskull replied:
Can you add a css class on the `<Tag>`? Otherwise, an invalid prop is being exposed to the user.

eriklharper reviewed (approved):
👍

macandcheese commented:
Ok, updated.. @asangma can you check to make sure this is still looking good 🕺  - I know there were a lot of css combos for the button alignment stuff that was added, I think this keeps them all in tact.

I also added a check for whitespace where it wouldn't be intended, for example -  on cards where line breaks existed between button tags in the html, that would now have unintentionally been interpreted as slotted content with this PR (as Screener picked up!).

driskull commented on src/components/calcite-button/calcite-button.tsx:
```diff
@@ -137,18 +137,21 @@ export class CalciteButton {
+      <Host dir={dir}>
         <Tag
           {...attributes}
+          {...className}
```
Can you just do:

```
<Tag
class={{[CSS. contentSlotted]: this.hasContent}}
```

?

macandcheese replied:
We could - it leaves an empty `class` attribute if `hasContent` is false though:
![Screen Shot 2021-01-07 at 5 06 30 PM](https://user-images.githubusercontent.com/4733155/103962059-d0d3db00-510a-11eb-82c4-c8d3ff79e6dd.png)

driskull replied:
what do you think @jcfranco? I think the empty class is ok but it's nitpicky.

paulcpederson replied:
I think the stencil way to do this is:

```jsx
<Tag class={{ [CSS.contentSlotted]: this.hasContent }}>
```

ie. conditionally add the `CSS.contentSlotted` class if `this.hasContent` is true.

driskull reviewed (approved):
👍

macandcheese commented:
@asangma could you check out this PR since it’s related to the content / alignment updates that were previously made? Thanks!

julio8a commented:
@asangma, following up the review for this. ^

julio8a commented:
The last ping for @asangma, otherwise, let's move forward with this PR.

asangma reviewed (approved):
LGTM.

julio8a commented:
Can we merge this? or is this something left?
